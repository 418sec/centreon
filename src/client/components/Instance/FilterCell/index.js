export { default as FilterCell } from './FilterCell'