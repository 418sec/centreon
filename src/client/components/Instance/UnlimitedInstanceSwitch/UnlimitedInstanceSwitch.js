import React from 'react'
import { withStyles } from 'material-ui/styles'
import { Switch } from 'material-ui'
import AllInclusive from 'material-ui-icons/AllInclusive'

const styles = {
  bar: {},
  checked: {
    '& + $bar': {
      backgroundColor: '#88b917',
    },
  },
  limitedIcon: {
    width: '14px',
    height: '14px',
    boxShadow: '0px 1px 3px 0px rgba(0, 0, 0, 0.2), 0px 1px 1px 0px rgba(0, 0, 0, 0.14), 0px 2px 1px -1px rgba(0, 0, 0, 0.12)',
    borderRadius: '50%',
    backgroundColor: '#e7e7e8',
    padding: '4px',
    fill: '#818285',
  },
  unlimitedIcon: {
    width: '14px',
    height: '14px',
    boxShadow: '0px 1px 3px 0px rgba(0, 0, 0, 0.2), 0px 1px 1px 0px rgba(0, 0, 0, 0.14), 0px 2px 1px -1px rgba(0, 0, 0, 0.12)',
    borderRadius: '50%',
    backgroundColor: '#88b917',
    padding: '4px',
    fill: '#FFFFFF',
  },
}

const UnlimitedInstanceSwitch = ({ classes, checked, onChange, instanceId }) => (
      <Switch
        classes={
          {
            checked: classes.checked,
            bar: classes.bar,
          }
        }
        id={'switch-' + instanceId}
        icon={<AllInclusive className={classes.limitedIcon} />}
        checked={checked}
        checkedIcon={<AllInclusive  className={classes.unlimitedIcon}/>}
        onChange={onChange}
      />
    )

export default withStyles(styles)(UnlimitedInstanceSwitch)