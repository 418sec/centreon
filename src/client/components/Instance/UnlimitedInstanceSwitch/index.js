export { default as UnlimitedInstanceSwitch } from './UnlimitedInstanceSwitchContainer'