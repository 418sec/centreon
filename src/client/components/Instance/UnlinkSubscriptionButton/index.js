export { default as UnlinkSubscriptionButton } from './UnlinkSubscriptionButtonContainer';