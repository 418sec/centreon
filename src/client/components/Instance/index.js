export { default as InstanceList } from './InstanceListContainer';