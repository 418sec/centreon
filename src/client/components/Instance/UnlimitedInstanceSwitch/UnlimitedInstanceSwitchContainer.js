import React, { Component } from 'react'
import { connect } from 'react-redux'
import UnlimitedInstanceSwitch from './UnlimitedInstanceSwitch'
import { openConfirmDialog, cleanConfirmDialog } from "Redux/confirmDialog/confirmDialogActions"
import { patchInstance } from 'Services/hubApi'

class UnlimitedInstanceSwitchContainer extends Component {

  onSwitch = (event, checked) => {
    const { openConfirmDialog, slug, serverName } = this.props
    const confirmDialog = {
      body: (checked ? 'Grant' : 'Revoke') + ' unlimited access to "' + serverName + '" ?',
      cancel: 'Cancel',
      confirm: 'Confirm'
    }
    openConfirmDialog(slug, confirmDialog)
  }

  componentWillReceiveProps = (nextProps) => {
    const { patchInstance, token, instanceId, checked, slug, allowedSlug } = this.props

    if (nextProps.allowAction === true && slug === allowedSlug) {
      let viewNotReleased = checked ? 0 : 1
      patchInstance(token, instanceId, {'view_not_released': viewNotReleased})
    }
  }

  render = () => {
    const { checked, instanceId } = this.props

    return (
      <UnlimitedInstanceSwitch
        instanceId={instanceId}
        checked={checked}
        onChange={this.onSwitch}
      />
    )
  }
}

const mapStateToProps = (store, ownProps) => {
  return {
    token: store.authReducer.token,
    serverName: store.instanceReducer.items[ownProps.instanceId].serverName,
    checked: store.instanceReducer.items[ownProps.instanceId].unlimitedAccess.unlimitedMode,
    allowAction: store.confirmDialogReducer.allowAction,
    allowedSlug: store.confirmDialogReducer.slug,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    openConfirmDialog: (slug, content) => {
      dispatch(openConfirmDialog(slug, content))
    },
    patchInstance: (token, instanceId, attributes) => {
      dispatch(patchInstance(token, instanceId, attributes))
    },
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(UnlimitedInstanceSwitchContainer)