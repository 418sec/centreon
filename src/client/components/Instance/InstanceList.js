import React, { Component } from 'react'
import {
  PagingState,
  FilteringState,
  CustomPaging
} from '@devexpress/dx-react-grid'
import {
  Grid,
  Table,
  TableHeaderRow,
  TableFilterRow,
  PagingPanel
} from '@devexpress/dx-react-grid-material-ui'
import PropTypes from 'prop-types'
import CheckCircle from 'material-ui-icons/CheckCircle'
import { withStyles } from 'material-ui/styles'
import { TableRow, TableCell } from 'material-ui'
import { UnlimitedInstanceSwitch } from 'Components/Instance/UnlimitedInstanceSwitch'
import { UnlinkSubscriptionButton } from 'Components/Instance/UnlinkSubscriptionButton'
import { FilterCell } from 'Components/Instance/FilterCell'

const styles = theme => ({
  root: {
    padding: '0 15px',
    overflow: 'auto'
  },
  HeaderRows: {
    position: 'sticky',
    zIndex: '1',
    top: '0'
  },
  TableCell: {
    wordWrap: 'break-word',
    padding: '4px 14px',
  },
  check: {
    fill: '#88B917'
  },
  checkInactive: {
    fill: '#d1d2d4'
  },
  dialog: {
    width: '80%',
    maxHeight: 435,
  },
})

class InstanceList extends Component {
  constructor (props) {
    super(props)

    this.state = {
      filters: [],
    }
  }

  tableRowTemplate = ({children, row}) => {
    return (
      <TableRow id={row.companyName.replace(' ', '-')}>
        {children}
      </TableRow>
    )
  }

  render = () => {
    const { columns, classes, rows, itemsPerPage, totalItems, currentPage } = this.props

    return (
      <div id="instanceList" className={classes.root}>
        <Grid
          columns={columns}
          rows={rows}
        >
          <FilteringState
            onFiltersChange={this.props.changeFilters}
          />
          <PagingState
            currentPage={currentPage}
            onCurrentPageChange={this.props.changeCurrentPage}
            pageSize={itemsPerPage}
          />
          <CustomPaging
            totalCount={totalItems}
          />
          <Table
            rowComponent={this.tableRowTemplate}
            cellComponent={({ column, value }) => {
              if (column.name === 'subscriptionState') {
                if (value.state) {
                  return (
                    <TableCell>
                      <UnlinkSubscriptionButton
                        instanceId={value.instanceId}
                        slug={value.slug}
                      />
                    </TableCell>
                  )
                } else {
                  return (
                    <TableCell>
                      <CheckCircle className={classes.checkInactive} />
                    </TableCell>
                  )
                }
              }
              if (column.name === 'unlimitedAccess') {
                return (
                  <TableCell className={classes.TableCell}>
                    <UnlimitedInstanceSwitch
                      instanceId={value.instanceId}
                      slug={value.slug}
                    />
                  </TableCell>
                )
              }
              return (
                <TableCell className={classes.TableCell}>{value}</TableCell>
              )
            }}
          />
          <TableHeaderRow />
          <PagingPanel />
          <TableFilterRow
            cellComponent={FilterCell}
          />
        </Grid>
      </div>
    )
  }
}

/**
 * Instance list Prototypes
 */
InstanceList.propTypes = {
  /**
   * Instance list css classes
   * @param {Object} classes
   */
  classes: PropTypes.object.isRequired,
}

InstanceList.defaultProps = {
  currentPage: 0
}

export default withStyles(styles)(InstanceList)