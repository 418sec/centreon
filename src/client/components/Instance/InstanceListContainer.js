import React, { Component } from 'react'
import InstanceList from './InstanceList'
import { connect } from 'react-redux'
import { getInstances } from 'Services/hubApi'
import { cleanHeader, replaceTitle} from 'Redux/header/headerActions'

export class InstanceListContainer extends Component {

  constructor (props) {
    super(props)

    this.state = {
      columns: [
        {name: 'companyName', title: 'Company name'},
        {name: 'fingerprint', title: 'Fingerprint'},
        {name: 'serverName', title: 'Server name'},
        {name: 'subscriptionState', title: 'Subscription state'},
        {name: 'unlimitedAccess', title: 'Unlimited access'},
      ],
      currentPage: 0,
      itemsPerPage: 12,
      filters: []
    }
  }

  componentDidMount = () => {
    const { cleanHeader, replaceHeaderTitle } = this.props
    cleanHeader()
    replaceHeaderTitle('Instances')

    this.getData()
  }

  changeFilters = (filters) => {
    this.setState(
      {
        filters: filters,
        currentPage: 0
      },
      this.getData
    )
  }

  changeCurrentPage = (currentPage) => {
    this.setState(
      {currentPage: currentPage},
      this.getData
    )
  }

  /**
   * example : page[number]=1&page[size]=30
   * @returns {string|*}
   */
  getPagination = () => {
    const { currentPage, itemsPerPage } = this.state
    return `?page[number]=${currentPage + 1}&page[size]=${itemsPerPage}`
  }

  /**
   * example : filter[instanceName]=centreon&filter[fingerprint]=abc
   * @returns {string|*}
   */
  getFilters = () => {
    const { filters } = this.state;

    return filters.reduce((acc, { columnName, value }) => {
      acc.push(`filter[${columnName}]=${value}`);
      return acc;
    }, []).join('&');
  }

  getData = () => {
    const { token } = this.props
    this.props.displayInstances(this.getPagination(), this.getFilters(), token)
  }

  render = () => {
    const { columns, currentPage } = this.state
    const { items, itemsPerPage, totalPages, totalItems } = this.props

    const rows = items ?
      Object.keys(items).reduce((acc, key) => {
        acc.push(items[key])
        return acc
      }, [])
      : []

    return (
      <InstanceList
        columns={columns}
        rows={rows}
        itemsPerPage={itemsPerPage}
        totalPages={totalPages}
        totalItems={totalItems}
        currentPage={currentPage}
        changeFilters={this.changeFilters}
        changeCurrentPage={this.changeCurrentPage}
      />
    )
  }
}

const mapStateToProps = (store) => {
  return {
    token: store.authReducer.token,
    items: store.instanceReducer.items,
    totalPages: store.instanceReducer.totalPages,
    totalItems: store.instanceReducer.totalItems,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    cleanHeader: () => {
      return dispatch(cleanHeader())
    },
    replaceHeaderTitle: (title) => {
      return dispatch(replaceTitle(title))
    },
    displayInstances: (pagination, filters, token) => {
      return dispatch(getInstances(pagination, filters, token))
    },
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(InstanceListContainer)