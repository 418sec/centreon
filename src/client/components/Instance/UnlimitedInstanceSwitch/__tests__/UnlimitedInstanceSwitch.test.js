/* eslint-env jest */
import React from 'react'
import Enzyme from 'enzyme'
import { mount, shallow, render } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import { UnlimitedInstanceSwitch } from 'Components/Instance/UnlimitedInstanceSwitch'
import { ConfirmDialog } from 'Components/ConfirmDialog'
import configureStore from 'redux-mock-store'

Enzyme.configure({ adapter: new Adapter() })

describe('UnlimitedInstanceSwitch', () => {
  const mockStore = configureStore()
  const store = mockStore({
    authReducer: {
      token: ''
    },
    confirmDialogReducer: {
      allowAction: false,
      slug: null,
      open: true,
      content: {
        body: 'message',
        cancel: 'cancel',
        confirm: 'confirm',
      },
    },
    instanceReducer: {
      items: {
        "1": {
          unlimitedAccess: {
            instanceId: "1",
            serverName: "centreon",
            slug:"instance-unlimited-1",
            unlimitedMode: true
          }
        }
      }
    }
  })

  it('Render button switch correctly', () => {

    const UnlimitedInstanceSwitchContainer = shallow(
      <UnlimitedInstanceSwitch
        store={store}
        instanceId="1"
      />
    )
    expect(toJson(UnlimitedInstanceSwitchContainer)).toMatchSnapshot()
  })

  it('Opens dialog on switch', () => {

    const UnlimitedInstanceSwitchContainer = mount(
      <UnlimitedInstanceSwitch
        store={store}
        instanceId="1"
      />
    )

    UnlimitedInstanceSwitchContainer
      .find('span#switch-1 input[type="checkbox"]')
      .simulate('change')

    const actions = store.getActions()

    expect(actions[0].type).toEqual('OPEN_CONFIRM_DIALOG')
  })

  it('Switch correctly on confirm', () => {
    const allowedSlug = "instance-unlimited-1"

    const UnlimitedInstanceSwitchContainer = mount(
      <UnlimitedInstanceSwitch
        store={store}
        instanceId="1"
        checked={0}
        allowAction={false}
        slug=""
        allowedSlug=""
      />
    )

    UnlimitedInstanceSwitchContainer
      .setProps({
        allowAction: true,
        checked: 1,
        allowedSlug: allowedSlug,
        slug: allowedSlug
      })
  })
})