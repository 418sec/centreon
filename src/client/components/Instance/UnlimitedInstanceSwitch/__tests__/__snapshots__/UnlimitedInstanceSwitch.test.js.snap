// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnlimitedInstanceSwitch Render button switch correctly 1`] = `
<UnlimitedInstanceSwitchContainer
  allowAction={false}
  allowedSlug={null}
  checked={true}
  instanceId="1"
  openConfirmDialog={[Function]}
  patchInstance={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  token=""
/>
`;
